Sporadic failures of Parquet `segarray_write` test with gasnet

**Describe the bug**
This failure:
```
        # bool with empty segments
        a = [0, 1, 1]
        c = [1, 0]
        s = ak.SegArray.from_parts(ak.array([0, 0, len(a), len(a), len(a), len(a) + len(c)]), ak.array(a + c,
                                   dtype=ak.bool))
        with tempfile.TemporaryDirectory(dir=ParquetTest.par_test_base_tmp) as tmp_dirname:
            s.to_parquet(f"{tmp_dirname}/int_test")

            rd_data = ak.read_parquet(f"{tmp_dirname}/int_test*")
            for i in range(6):
>               self.assertListEqual(s[i].tolist(), rd_data[i].tolist())
E               AssertionError: Lists differ: [False, True, True] != []
E
E               First list contains 3 additional elements.
E               First extra element 0:
E               False
E
E               - [False, True, True]
E               + []

tests/parquet_test.py:483: AssertionError
```
popped up on nightly testing. I am able to reproduce this issue on 16 locales, but it appears to only show up about once in every 100 trials.

**To Reproduce**

- Edit `pytest.ini` to run with 4 or more locales (not sure if this is necessary, but I couldn't reproduce with 2 locales, but could with 16; fairly confident anything 4 or over would work based on the type of failure and what we've seen here in the past)
- Modify `test_segarray_write` to run a large number of trials (I seem to catch this about 1 in every 100 runs)
- Run `python3 -m pytest tests/parquet_test.py -k "test_segarray_write"` to run just this test

--- a/src/ArrowFunctions.cpp
+++ b/src/ArrowFunctions.cpp
@@ -328,6 +328,7 @@ int64_t cpp_getListColumnSize(const char* filename, const char* colname, void* c
             (void)int_reader->ReadBatch(1, &definition_level, &rep_lvl, &value, &values_read);
             if (values_read == 0 || (!first && rep_lvl == 0)) {
               seg_sizes[i] = seg_size;
+              std::cout << "\n\ni: "<<i<<"\nseg_sizes[i]: "<<seg_sizes[i]<<"\n\n";
               i++;
               seg_size = 0;
             }
@@ -338,8 +339,9 @@ int64_t cpp_getListColumnSize(const char* filename, const char* colname, void* c
                 first = false;
               }
             }
-            if (!int_reader->HasNext()){
+            if (values_read != 0 && !int_reader->HasNext()){
               seg_sizes[i] = seg_size;
+              std::cout << "\n\ni: "<<i<<"\nseg_sizes[i]: "<<seg_sizes[i]<<"\n\n";
             }
           }
         } else if(lty == ARROWINT32 || lty == ARROWUINT32) {
@@ -557,13 +559,13 @@ int cpp_readListColumnByName(const char* filename, void* chpl_arr, const char* c
           auto chpl_ptr = (int64_t*)chpl_arr;
           parquet::Int64Reader* reader =
             static_cast<parquet::Int64Reader*>(column_reader.get());
-          startIdx -= reader->Skip(startIdx);
+          // startIdx -= reader->Skip(startIdx);
 
           while (reader->HasNext() && i < numElems) {
             if((numElems - i) < batchSize)
               batchSize = numElems - i;
             (void)reader->ReadBatch(batchSize, nullptr, nullptr, &chpl_ptr[i], &values_read);
-            i+=values_read;
+            i+=1;
           }
         } else if(lty == ARROWINT32 || lty == ARROWUINT32) {
           auto chpl_ptr = (int64_t*)chpl_arr;
@@ -621,7 +623,6 @@ int cpp_readListColumnByName(const char* filename, void* chpl_arr, const char* c
           while (reader->HasNext() && i < numElems) {
             if((numElems - i) < batchSize) // adjust batchSize if needed
               batchSize = numElems - i;
-            
             // unlike the base case, the list stores NaN values and is able to read them out. 
             // This prevents the need for any additional processing
             // setting nullptr for def and rep levels allows use to ignore the empty segments since we only care about values here.
@@ -633,6 +634,7 @@ int cpp_readListColumnByName(const char* filename, void* chpl_arr, const char* c
               chpl_ptr[i+j] = (double)tmpArr[j];
             }
             i += values_read;
+            delete[] tmpArr;
           }
         } else if(lty == ARROWDOUBLE) {
           auto chpl_ptr = (double*)chpl_arr;
@@ -643,7 +645,6 @@ int cpp_readListColumnByName(const char* filename, void* chpl_arr, const char* c
           while (reader->HasNext() && i < numElems) {
             if((numElems - i) < batchSize) // adjust batchSize if needed
               batchSize = numElems - i;
-              
             // unlike the base case, the list stores NaN values and is able to read them out. 
             // This prevents the need for any additional processing
             // setting nullptr for def and rep levels allows use to ignore the empty segments since we only care about values here.
@@ -783,6 +784,8 @@ int cpp_readColumnByName(const char* filename, void* chpl_arr, const char* colna
             }
             i++;
           }
+          delete[] def_lvl;
+          delete[] rep_lvl;
         }
       } else if(ty == ARROWDOUBLE) {
         auto chpl_ptr = (double*)chpl_arr;
@@ -812,6 +815,9 @@ int cpp_readColumnByName(const char* filename, void* chpl_arr, const char* colna
             }
             i++;
           }
+          delete[] tmpArr;
+          delete[] def_lvl;
+          delete[] rep_lvl;
         }
       }
     }
@@ -963,6 +969,8 @@ int cpp_writeMultiColToParquet(const char* filename, void* column_names,
                 }
                 int64_t valIdx = offset_ptr[offIdx];
                 writer->WriteBatch(segSize, def_lvl, rep_lvl, &data_ptr[valIdx]);
+                delete[] def_lvl;
+                delete[] rep_lvl;
               }
               else {
                 // empty segment denoted by null value that is not repeated (first of segment) defined at the list level (1)
@@ -970,6 +978,8 @@ int cpp_writeMultiColToParquet(const char* filename, void* column_names,
                 int16_t* def_lvl = new int16_t[segSize] { 1 };
                 int16_t* rep_lvl = new int16_t[segSize] { 0 };
                 writer->WriteBatch(segSize, def_lvl, rep_lvl, nullptr);
+                delete[] def_lvl;
+                delete[] rep_lvl;
               }
               offIdx++;
               count++;
@@ -1013,6 +1023,8 @@ int cpp_writeMultiColToParquet(const char* filename, void* column_names,
                 }
                 int64_t valIdx = offset_ptr[offIdx];
                 writer->WriteBatch(segSize, def_lvl, rep_lvl, &data_ptr[valIdx]);
+                delete[] def_lvl;
+                delete[] rep_lvl;
               }
               else {
                 // empty segment denoted by null value that is not repeated (first of segment) defined at the list level (1)
@@ -1020,6 +1032,8 @@ int cpp_writeMultiColToParquet(const char* filename, void* column_names,
                 int16_t* def_lvl = new int16_t[segSize] { 1 };
                 int16_t* rep_lvl = new int16_t[segSize] { 0 };
                 writer->WriteBatch(segSize, def_lvl, rep_lvl, nullptr);
+                delete[] def_lvl;
+                delete[] rep_lvl;
               }
               offIdx++;
               count++;
@@ -1063,6 +1077,8 @@ int cpp_writeMultiColToParquet(const char* filename, void* column_names,
                 }
                 int64_t valIdx = offset_ptr[offIdx];
                 writer->WriteBatch(segSize, def_lvl, rep_lvl, &data_ptr[valIdx]);
+                delete[] def_lvl;
+                delete[] rep_lvl;
               }
               else {
                 // empty segment denoted by null value that is not repeated (first of segment) defined at the list level (1)
@@ -1070,6 +1086,8 @@ int cpp_writeMultiColToParquet(const char* filename, void* column_names,
                 int16_t* def_lvl = new int16_t[segSize] { 1 };
                 int16_t* rep_lvl = new int16_t[segSize] { 0 };
                 writer->WriteBatch(segSize, def_lvl, rep_lvl, nullptr);
+                delete[] def_lvl;
+                delete[] rep_lvl;
               }
               offIdx++;
               count++;
@@ -1512,9 +1530,7 @@ int cpp_writeListColumnToParquet(const char* filename, void* chpl_segs, void* ch
     int64_t valIdx = 0; // index into chpl_arr
     int64_t segIdx = 0; // index into offsets
 
-    if(dtype == ARROWINT64 || dtype == ARROWUINT64) {
-      auto chpl_ptr = (int64_t*)chpl_arr;
-      
+    if(dtype == ARROWINT64 || dtype == ARROWUINT64) {     
       while(numLeft > 0) { // write all local values to the file
         parquet::RowGroupWriter* rg_writer = file_writer->AppendRowGroup();
         parquet::Int64Writer* writer =
@@ -1523,23 +1539,25 @@ int cpp_writeListColumnToParquet(const char* filename, void* chpl_segs, void* ch
         while (numLeft > 0 && count < rowGroupSize) { // ensures rowGroupSize maintained
           int64_t batchSize = segments[segIdx+1] - segments[segIdx];
           if (batchSize > 0) {
-            int16_t* def_lvl = new int16_t[batchSize] { 3 };
+            auto chpl_ptr = (int64_t*)chpl_arr;
+            int16_t* def_lvl = new int16_t[batchSize] { 3 }; // all values defined at the item level (3)
             int16_t* rep_lvl = new int16_t[batchSize] { 0 };
             for (int64_t x = 0; x < batchSize; x++){
               // if the value is first in the segment rep_lvl = 0, otherwise 1
-              // all values defined at the item level (3)
               rep_lvl[x] = (x == 0) ? 0 : 1;
               def_lvl[x] = 3;
             }
             writer->WriteBatch(batchSize, def_lvl, rep_lvl, &chpl_ptr[valIdx]);
             valIdx += batchSize;
+            delete[] def_lvl;
+            delete[] rep_lvl;
           }
           else {
             // empty segment denoted by null value that is not repeated (first of segment) defined at the list level (1)
             batchSize = 1; // even though segment is length=0, write null to hold the empty segment
-            int16_t* def_lvl = new int16_t[batchSize] { 1 };
-            int16_t* rep_lvl = new int16_t[batchSize] { 0 };
-            writer->WriteBatch(batchSize, def_lvl, rep_lvl, nullptr);
+            int16_t def_lvl = 1;
+            int16_t rep_lvl = 0;
+            writer->WriteBatch(batchSize, &def_lvl, &rep_lvl, nullptr);
           }
           count++;
           segIdx++;
@@ -1548,8 +1566,6 @@ int cpp_writeListColumnToParquet(const char* filename, void* chpl_segs, void* ch
       }
     }
     else if (dtype == ARROWBOOLEAN) {
-      auto chpl_ptr = (bool*)chpl_arr;
-      
       while(numLeft > 0) {
         parquet::RowGroupWriter* rg_writer = file_writer->AppendRowGroup();
         parquet::BoolWriter* writer =
@@ -1558,6 +1574,7 @@ int cpp_writeListColumnToParquet(const char* filename, void* chpl_segs, void* ch
         while (numLeft > 0 && count < rowGroupSize) {
           int64_t batchSize = segments[segIdx+1] - segments[segIdx];
           if (batchSize > 0) {
+            auto chpl_ptr = (bool*)chpl_arr;
             // if the value is first in the segment rep_lvl = 0, otherwise 1
             // all values defined at the item level (3)
             int16_t* def_lvl = new int16_t[batchSize] { 3 };
@@ -1568,13 +1585,15 @@ int cpp_writeListColumnToParquet(const char* filename, void* chpl_segs, void* ch
             }
             writer->WriteBatch(batchSize, def_lvl, rep_lvl, &chpl_ptr[valIdx]);
             valIdx += batchSize;
+            delete[] def_lvl;
+            delete[] rep_lvl;
           }
           else {
             // empty segment denoted by null value that is not repeated (first of segment) defined at the list level (1)
             batchSize = 1; // even though segment is length=0, write null to hold the empty segment
-            int16_t* def_lvl = new int16_t[batchSize] { 1 };
-            int16_t* rep_lvl = new int16_t[batchSize] { 0 };
-            writer->WriteBatch(batchSize, def_lvl, rep_lvl, nullptr);
+            int16_t def_lvl = 1;
+            int16_t rep_lvl = 0;
+            writer->WriteBatch(batchSize, &def_lvl, &rep_lvl, nullptr);
           }
           count++;
           segIdx++;
@@ -1582,9 +1601,7 @@ int cpp_writeListColumnToParquet(const char* filename, void* chpl_segs, void* ch
         }
       }
     }
-    else if (dtype == ARROWDOUBLE) {
-      auto chpl_ptr = (double*)chpl_arr;
-      
+    else if (dtype == ARROWDOUBLE) {      
       while(numLeft > 0) {
         parquet::RowGroupWriter* rg_writer = file_writer->AppendRowGroup();
         parquet::DoubleWriter* writer =
@@ -1593,6 +1610,7 @@ int cpp_writeListColumnToParquet(const char* filename, void* chpl_segs, void* ch
         while (numLeft > 0 && count < rowGroupSize) {
           int64_t batchSize = segments[segIdx+1] - segments[segIdx];
           if (batchSize > 0) {
+            auto chpl_ptr = (double*)chpl_arr;
             // if the value is first in the segment rep_lvl = 0, otherwise 1
             // all values defined at the item level (3)
             int16_t* def_lvl = new int16_t[batchSize] { 3 };
@@ -1603,13 +1621,15 @@ int cpp_writeListColumnToParquet(const char* filename, void* chpl_segs, void* ch
             }
             writer->WriteBatch(batchSize, def_lvl, rep_lvl, &chpl_ptr[valIdx]);
             valIdx += batchSize;
+            delete[] def_lvl;
+            delete[] rep_lvl;
           }
           else {
             // empty segment denoted by null value that is not repeated (first of segment) defined at the list level (1)
             batchSize = 1; // even though segment is length=0, write null to hold the empty segment
-            int16_t* def_lvl = new int16_t[batchSize] { 1 };
-            int16_t* rep_lvl = new int16_t[batchSize] { 0 };
-            writer->WriteBatch(batchSize, def_lvl, rep_lvl, nullptr);
+            int16_t def_lvl = 1;
+            int16_t rep_lvl = 0;
+            writer->WriteBatch(batchSize, &def_lvl, &rep_lvl, nullptr);
           }
           count++;
           segIdx++;

--- a/src/ArrowFunctions.cpp
+++ b/src/ArrowFunctions.cpp
@@ -328,7 +328,6 @@ int64_t cpp_getListColumnSize(const char* filename, const char* colname, void* c
             (void)int_reader->ReadBatch(1, &definition_level, &rep_lvl, &value, &values_read);
             if (values_read == 0 || (!first && rep_lvl == 0)) {
               seg_sizes[i] = seg_size;
-              std::cout << "\n\ni: "<<i<<"\nseg_sizes[i]: "<<seg_sizes[i]<<"\n\n";
               i++;
               seg_size = 0;
             }
@@ -341,7 +340,6 @@ int64_t cpp_getListColumnSize(const char* filename, const char* colname, void* c
             }
             if (values_read != 0 && !int_reader->HasNext()){
               seg_sizes[i] = seg_size;
-              std::cout << "\n\ni: "<<i<<"\nseg_sizes[i]: "<<seg_sizes[i]<<"\n\n";
             }
           }
         } else if(lty == ARROWINT32 || lty == ARROWUINT32) {
@@ -363,7 +361,7 @@ int64_t cpp_getListColumnSize(const char* filename, const char* colname, void* c
                 first = false;
               }
             }
-            if (!int_reader->HasNext()){
+            if (values_read != 0 && !int_reader->HasNext()){
               seg_sizes[i] = seg_size;
             }
           }
@@ -386,7 +384,7 @@ int64_t cpp_getListColumnSize(const char* filename, const char* colname, void* c
                 first = false;
               }
             }
-            if (!reader->HasNext()){
+            if (values_read != 0 && !reader->HasNext()){
               seg_sizes[i] = seg_size;
             }
           }
@@ -409,7 +407,7 @@ int64_t cpp_getListColumnSize(const char* filename, const char* colname, void* c
                 first = false;
               }
             }
-            if (!bool_reader->HasNext()){
+            if (values_read != 0 && !bool_reader->HasNext()){
               seg_sizes[i] = seg_size;
             }
           }
@@ -434,7 +432,7 @@ int64_t cpp_getListColumnSize(const char* filename, const char* colname, void* c
                 first = false;
               }
             }
-            if (!float_reader->HasNext()){
+            if (values_read != 0 && !float_reader->HasNext()){
               seg_sizes[i] = seg_size;
             }
           }
@@ -458,7 +456,7 @@ int64_t cpp_getListColumnSize(const char* filename, const char* colname, void* c
                 first = false;
               }
             }
-            if (!dbl_reader->HasNext()){
+            if (values_read != 0 && !dbl_reader->HasNext()){
               seg_sizes[i] = seg_size;
             }
           }
@@ -565,7 +563,7 @@ int cpp_readListColumnByName(const char* filename, void* chpl_arr, const char* c
             if((numElems - i) < batchSize)
               batchSize = numElems - i;
             (void)reader->ReadBatch(batchSize, nullptr, nullptr, &chpl_ptr[i], &values_read);
-            i+=1;
+            i+=values_read;
           }
         } else if(lty == ARROWINT32 || lty == ARROWUINT32) {
           auto chpl_ptr = (int64_t*)chpl_arr;
@@ -1248,6 +1246,10 @@ int cpp_writeColumnToParquet(const char* filename, void* chpl_arr,
     int64_t i = 0;
     int64_t numLeft = numelems;
 
+    if (chpl_arr == NULL) {
+      return 0;
+    }
+
     if(dtype == ARROWINT64 || dtype == ARROWUINT64) {
       auto chpl_ptr = (int64_t*)chpl_arr;
       while(numLeft > 0) {
@@ -1772,6 +1774,9 @@ int cpp_appendColumnToParquet(const char* filename, void* chpl_arr,
                               int64_t dtype, int64_t compression,
                               char** errMsg) {
   try {
+    if (chpl_arr == NULL){
+      return 0;
+    }
     std::shared_ptr<arrow::io::ReadableFile> infile;
     ARROWRESULT_OK(arrow::io::ReadableFile::Open(filename, arrow::default_memory_pool()),
                    infile);

--- a/src/ArrowFunctions.cpp
+++ b/src/ArrowFunctions.cpp
@@ -557,7 +557,6 @@ int cpp_readListColumnByName(const char* filename, void* chpl_arr, const char* c
           auto chpl_ptr = (int64_t*)chpl_arr;
           parquet::Int64Reader* reader =
             static_cast<parquet::Int64Reader*>(column_reader.get());
-          // startIdx -= reader->Skip(startIdx);
 
           while (reader->HasNext() && i < numElems) {
             if((numElems - i) < batchSize)
@@ -569,7 +568,6 @@ int cpp_readListColumnByName(const char* filename, void* chpl_arr, const char* c
           auto chpl_ptr = (int64_t*)chpl_arr;
           parquet::Int32Reader* reader =
             static_cast<parquet::Int32Reader*>(column_reader.get());
-          startIdx -= reader->Skip(startIdx);
 
           int32_t* tmpArr = (int32_t*)malloc(batchSize * sizeof(int32_t));
           while (reader->HasNext() && i < numElems) {
@@ -604,7 +602,6 @@ int cpp_readListColumnByName(const char* filename, void* chpl_arr, const char* c
           auto chpl_ptr = (bool*)chpl_arr;
           parquet::BoolReader* reader =
             static_cast<parquet::BoolReader*>(column_reader.get());
-          startIdx -= reader->Skip(startIdx);
 
           while (reader->HasNext() && i < numElems) {
             if((numElems - i) < batchSize)
@@ -616,7 +613,6 @@ int cpp_readListColumnByName(const char* filename, void* chpl_arr, const char* c
           auto chpl_ptr = (double*)chpl_arr;
           parquet::FloatReader* reader =
             static_cast<parquet::FloatReader*>(column_reader.get());
-          startIdx -= reader->Skip(startIdx);
           
           while (reader->HasNext() && i < numElems) {
             if((numElems - i) < batchSize) // adjust batchSize if needed
@@ -638,7 +634,6 @@ int cpp_readListColumnByName(const char* filename, void* chpl_arr, const char* c
           auto chpl_ptr = (double*)chpl_arr;
           parquet::DoubleReader* reader =
             static_cast<parquet::DoubleReader*>(column_reader.get());
-          startIdx -= reader->Skip(startIdx);
 
           while (reader->HasNext() && i < numElems) {
             if((numElems - i) < batchSize) // adjust batchSize if needed
@@ -973,11 +968,9 @@ int cpp_writeMultiColToParquet(const char* filename, void* column_names,
               else {
                 // empty segment denoted by null value that is not repeated (first of segment) defined at the list level (1)
                 segSize = 1; // even though segment is length=0, write null to hold the empty segment
-                int16_t* def_lvl = new int16_t[segSize] { 1 };
-                int16_t* rep_lvl = new int16_t[segSize] { 0 };
-                writer->WriteBatch(segSize, def_lvl, rep_lvl, nullptr);
-                delete[] def_lvl;
-                delete[] rep_lvl;
+                int16_t def_lvl = 1;
+                int16_t rep_lvl = 0;
+                writer->WriteBatch(segSize, &def_lvl, &rep_lvl, nullptr);
               }
               offIdx++;
               count++;
@@ -1027,11 +1020,9 @@ int cpp_writeMultiColToParquet(const char* filename, void* column_names,
               else {
                 // empty segment denoted by null value that is not repeated (first of segment) defined at the list level (1)
                 segSize = 1; // even though segment is length=0, write null to hold the empty segment
-                int16_t* def_lvl = new int16_t[segSize] { 1 };
-                int16_t* rep_lvl = new int16_t[segSize] { 0 };
-                writer->WriteBatch(segSize, def_lvl, rep_lvl, nullptr);
-                delete[] def_lvl;
-                delete[] rep_lvl;
+                int16_t def_lvl = 1;
+                int16_t rep_lvl = 0;
+                writer->WriteBatch(segSize, &def_lvl, &rep_lvl, nullptr);
               }
               offIdx++;
               count++;
@@ -1081,11 +1072,9 @@ int cpp_writeMultiColToParquet(const char* filename, void* column_names,
               else {
                 // empty segment denoted by null value that is not repeated (first of segment) defined at the list level (1)
                 segSize = 1; // even though segment is length=0, write null to hold the empty segment
-                int16_t* def_lvl = new int16_t[segSize] { 1 };
-                int16_t* rep_lvl = new int16_t[segSize] { 0 };
-                writer->WriteBatch(segSize, def_lvl, rep_lvl, nullptr);
-                delete[] def_lvl;
-                delete[] rep_lvl;
+                int16_t def_lvl = 1;
+                int16_t rep_lvl =0;
+                writer->WriteBatch(segSize, &def_lvl, &rep_lvl, nullptr);
               }
               offIdx++;
               count++;

--- a/src/ArrowFunctions.cpp
+++ b/src/ArrowFunctions.cpp
@@ -1236,6 +1236,7 @@ int cpp_writeColumnToParquet(const char* filename, void* chpl_arr,
     int64_t numLeft = numelems;
 
     if (chpl_arr == NULL) {
+      // early out to prevent bad memory access
       return 0;
     }
 
@@ -1764,6 +1765,7 @@ int cpp_appendColumnToParquet(const char* filename, void* chpl_arr,
                               char** errMsg) {
   try {
     if (chpl_arr == NULL){
+      // early out to prevent bad memory access
       return 0;
     }
     std::shared_ptr<arrow::io::ReadableFile> infile;

--- a/src/ParquetMsg.chpl
+++ b/src/ParquetMsg.chpl
@@ -195,15 +195,18 @@ module ParquetMsg {
     var fileOffsets = (+ scan sizes) - sizes;
     var segmentOffsets = (+ scan rows_per_file) - rows_per_file;
     
-    coforall loc in A.targetLocales() do on loc {
+    coforall loc in A.targetLocales() do on loc { // TODO return to coforall
       var locFiles = filenames;
       var locFiledoms = subdoms;
       var locOffsets = fileOffsets; // value count offset
       var locSegOffsets = segmentOffsets; // indicates which segment index is first for the file
+      var pqErr = new parquetErrorMsg();
 
-      forall (s, off, filedom, filename) in zip(locSegOffsets, locOffsets, locFiledoms, locFiles) {
+      // forall (s, off, filedom, filename) in zip(locSegOffsets, locOffsets, locFiledoms, locFiles) {
+      for (s, off, filedom, filename) in zip(locSegOffsets, locOffsets, locFiledoms, locFiles) {
         for locdom in A.localSubdomains() {
           const intersection = domain_intersection(locdom, filedom);
+          
           if intersection.size > 0 {
             var pqErr = new parquetErrorMsg();
             if ty == ArrowTypes.stringArr {
@@ -216,12 +219,19 @@ module ParquetMsg {
               A[filedom] = col;
             }
             else {
-              var shift = computeEmptySegs(seg_sizes, offsets, s, intersection, off); // compute the shift to account for any empty segments in the file before the current section.
-              if c_readListColumnByName(filename.localize().c_str(), c_ptrTo(A[intersection.low]),
-                                    dsetname.localize().c_str(), intersection.size, (intersection.low - off) + shift,
+              // var shift = computeEmptySegs(seg_sizes, offsets, s, intersection, off); // compute the shift to account for any empty segments in the file before the current section.
+              var col: [filedom] t;
+              // if c_readListColumnByName(filename.localize().c_str(), c_ptrTo(A[intersection.low]),
+              //                       dsetname.localize().c_str(), intersection.size, (intersection.low - off) + shift,
+              //                       batchSize, c_ptrTo(pqErr.errMsg)) == ARROWERROR {
+              //   pqErr.parquetError(getLineNumber(), getRoutineName(), getModuleName());
+              // }
+              if c_readListColumnByName(filename.localize().c_str(), c_ptrTo(col),
+                                    dsetname.localize().c_str(), filedom.size, 0,
                                     batchSize, c_ptrTo(pqErr.errMsg)) == ARROWERROR {
                 pqErr.parquetError(getLineNumber(), getRoutineName(), getModuleName());
               }
+              A[filedom] = col;
             }
           }
         }
@@ -328,6 +338,7 @@ module ParquetMsg {
     
     if listSize == ARROWERROR then
       pqErr.parquetError(getLineNumber(), getRoutineName(), getModuleName());
+    writeln("\n\nListSize: %jt\nSeg_Sizes: %jt\n\n".format(listSize, seg_sizes));
     return listSize;
   }
   
@@ -665,7 +676,7 @@ module ParquetMsg {
       var entrySeg = new shared SymEntry((+ reduce listSizes), int);
       var byteSizes = calcStrSizesAndOffset(entrySeg.a, filenames, listSizes, dsetname);
       entrySeg.a = (+ scan entrySeg.a) - entrySeg.a;
-      
+
       var entryVal = new shared SymEntry((+ reduce byteSizes), uint(8));
       readListFilesByName(entryVal.a, sizes, seg_sizes, segments, filenames, byteSizes, dsetname, ty);
       var stringsEntry = assembleSegStringFromParts(entrySeg, entryVal, st);
@@ -753,23 +764,23 @@ module ParquetMsg {
     dsetnames = dsetlist;
 
     if filelist.size == 1 {
-        if filelist[0].strip().size == 0 {
-            var errorMsg = "filelist was empty.";
-            pqLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);
-            return new MsgTuple(errorMsg, MsgType.ERROR);
-        }
-        var tmp = glob(filelist[0]);
-        pqLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),
-                              "glob expanded %s to %i files".format(filelist[0], tmp.size));
-        if tmp.size == 0 {
-            var errorMsg = "The wildcarded filename %s either corresponds to files inaccessible to Arkouda or files of an invalid format".format(filelist[0]);
-            pqLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);
-            return new MsgTuple(errorMsg, MsgType.ERROR);
-        }
-        // Glob returns filenames in weird order. Sort for consistency
-        sort(tmp);
-        filedom = tmp.domain;
-        filenames = tmp;
+      if filelist[0].strip().size == 0 {
+          var errorMsg = "filelist was empty.";
+          pqLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);
+          return new MsgTuple(errorMsg, MsgType.ERROR);
+      }
+      var tmp = glob(filelist[0]);
+      pqLogger.debug(getModuleName(),getRoutineName(),getLineNumber(),
+                            "glob expanded %s to %i files".format(filelist[0], tmp.size));
+      if tmp.size == 0 {
+          var errorMsg = "The wildcarded filename %s either corresponds to files inaccessible to Arkouda or files of an invalid format".format(filelist[0]);
+          pqLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);
+          return new MsgTuple(errorMsg, MsgType.ERROR);
+      }
+      // Glob returns filenames in weird order. Sort for consistency
+      sort(tmp);
+      filedom = tmp.domain;
+      filenames = tmp;
     } else {
         filenames = filelist;
     }
@@ -999,7 +1010,12 @@ module ParquetMsg {
 
     var pqErr = new parquetErrorMsg();
 
-    if c_writeListColumnToParquet(filename.localize().c_str(), c_ptrTo(locOffsets), c_ptrTo(localVals),
+    var valPtr: c_void_ptr = nil;
+    if localVals.size != 0 {
+      valPtr = c_ptrTo(localVals);
+    }
+
+    if c_writeListColumnToParquet(filename.localize().c_str(), c_ptrTo(locOffsets), valPtr,
                                    dsetname.localize().c_str(), locOffsets.size-1, ROWGROUPS,
                                    c_dtype, compression, c_ptrTo(pqErr.errMsg)) == ARROWERROR {
         pqErr.parquetError(getLineNumber(), getRoutineName(), getModuleName());
@@ -1031,7 +1047,6 @@ module ParquetMsg {
     // pull values to the locale of the offset
     coforall (loc, idx) in zip(segments.targetLocales(), filenames.domain) with (ref olda) do on loc {
       const myFilename = filenames[idx];
-
       const locDom = segments.localSubdomain();
       var dims: [0..#1] int;
       dims[0] = locDom.size: int;

--- a/src/ParquetMsg.chpl
+++ b/src/ParquetMsg.chpl
@@ -195,44 +195,25 @@ module ParquetMsg {
     var fileOffsets = (+ scan sizes) - sizes;
     var segmentOffsets = (+ scan rows_per_file) - rows_per_file;
     
-    coforall loc in A.targetLocales() do on loc { // TODO return to coforall
+    coforall loc in A.targetLocales() do on loc {
       var locFiles = filenames;
       var locFiledoms = subdoms;
       var locOffsets = fileOffsets; // value count offset
       var locSegOffsets = segmentOffsets; // indicates which segment index is first for the file
-      var pqErr = new parquetErrorMsg();
 
-      // forall (s, off, filedom, filename) in zip(locSegOffsets, locOffsets, locFiledoms, locFiles) {
-      for (s, off, filedom, filename) in zip(locSegOffsets, locOffsets, locFiledoms, locFiles) {
+      forall (s, off, filedom, filename) in zip(locSegOffsets, locOffsets, locFiledoms, locFiles) {
         for locdom in A.localSubdomains() {
           const intersection = domain_intersection(locdom, filedom);
           
           if intersection.size > 0 {
             var pqErr = new parquetErrorMsg();
-            if ty == ArrowTypes.stringArr {
-              var col: [filedom] t;
-              if c_readListColumnByName(filename.localize().c_str(), c_ptrTo(col),
-                                  dsetname.localize().c_str(), intersection.size, 0,
+            var col: [filedom] t;
+            if c_readListColumnByName(filename.localize().c_str(), c_ptrTo(col),
+                                  dsetname.localize().c_str(), filedom.size, 0,
                                   batchSize, c_ptrTo(pqErr.errMsg)) == ARROWERROR {
-                pqErr.parquetError(getLineNumber(), getRoutineName(), getModuleName());
-              }
-              A[filedom] = col;
+              pqErr.parquetError(getLineNumber(), getRoutineName(), getModuleName());
             }
-            else {
-              // var shift = computeEmptySegs(seg_sizes, offsets, s, intersection, off); // compute the shift to account for any empty segments in the file before the current section.
-              var col: [filedom] t;
-              // if c_readListColumnByName(filename.localize().c_str(), c_ptrTo(A[intersection.low]),
-              //                       dsetname.localize().c_str(), intersection.size, (intersection.low - off) + shift,
-              //                       batchSize, c_ptrTo(pqErr.errMsg)) == ARROWERROR {
-              //   pqErr.parquetError(getLineNumber(), getRoutineName(), getModuleName());
-              // }
-              if c_readListColumnByName(filename.localize().c_str(), c_ptrTo(col),
-                                    dsetname.localize().c_str(), filedom.size, 0,
-                                    batchSize, c_ptrTo(pqErr.errMsg)) == ARROWERROR {
-                pqErr.parquetError(getLineNumber(), getRoutineName(), getModuleName());
-              }
-              A[filedom] = col;
-            }
+            A[filedom] = col;
           }
         }
       }

--- a/src/ParquetMsg.chpl
+++ b/src/ParquetMsg.chpl
@@ -338,7 +338,6 @@ module ParquetMsg {
     
     if listSize == ARROWERROR then
       pqErr.parquetError(getLineNumber(), getRoutineName(), getModuleName());
-    writeln("\n\nListSize: %jt\nSeg_Sizes: %jt\n\n".format(listSize, seg_sizes));
     return listSize;
   }
   
@@ -465,14 +464,18 @@ module ParquetMsg {
 
         var locDom = A.localSubdomain();
         var locArr = A[locDom];
+        var valPtr: c_void_ptr = nil;
+        if locArr.size != 0 {
+          valPtr = c_ptrTo(locArr);
+        }
         if mode == TRUNCATE || !filesExist {
-          if c_writeColumnToParquet(myFilename.localize().c_str(), c_ptrTo(locArr), 0,
+          if c_writeColumnToParquet(myFilename.localize().c_str(), valPtr, 0,
                                     dsetname.localize().c_str(), locDom.size, rowGroupSize,
                                     dtypeRep, compression, c_ptrTo(pqErr.errMsg)) == ARROWERROR {
             pqErr.parquetError(getLineNumber(), getRoutineName(), getModuleName());
           }
         } else {
-          if c_appendColumnToParquet(myFilename.localize().c_str(), c_ptrTo(locArr),
+          if c_appendColumnToParquet(myFilename.localize().c_str(), valPtr,
                                      dsetname.localize().c_str(), locDom.size,
                                      dtypeRep, compression, c_ptrTo(pqErr.errMsg)) == ARROWERROR {
             pqErr.parquetError(getLineNumber(), getRoutineName(), getModuleName());

--- a/tests/parquet_test.py
+++ b/tests/parquet_test.py
@@ -346,17 +346,19 @@ class ParquetTest(ArkoudaTest):
             rd_data = ak.read_parquet(f"{tmp_dirname}/int_test*")
             for i in range(3):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
+        print("Finished Int")
 
         # integer with empty segments
         a = [0, 1, 2]
         c = [15, 21]
         s = ak.SegArray(ak.array([0, 0, len(a), len(a), len(a), len(a) + len(c)]), ak.array(a + c))
         with tempfile.TemporaryDirectory(dir=ParquetTest.par_test_base_tmp) as tmp_dirname:
-            s.to_parquet(f"{tmp_dirname}/int_test")
+            s.to_parquet(f"{tmp_dirname}/int_test_empty")
 
-            rd_data = ak.read_parquet(f"{tmp_dirname}/int_test*")
+            rd_data = ak.read_parquet(f"{tmp_dirname}/int_test_empty*")
             for i in range(6):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
+        print("Finished Int Empty")
 
         # uint test
         a = [0, 1, 2]
@@ -364,11 +366,13 @@ class ParquetTest(ArkoudaTest):
         c = [15, 21]
         s = ak.SegArray(ak.array([0, len(a), len(a) + len(b)]), ak.array(a + b + c, dtype=ak.uint64))
         with tempfile.TemporaryDirectory(dir=ParquetTest.par_test_base_tmp) as tmp_dirname:
-            s.to_parquet(f"{tmp_dirname}/int_test")
+            s.to_parquet(f"{tmp_dirname}/uint_test")
 
-            rd_data = ak.read_parquet(f"{tmp_dirname}/int_test*")
+            rd_data = ak.read_parquet(f"{tmp_dirname}/uint_test*")
             for i in range(3):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
+
+        print("Finished UInt")
 
         # uint with empty segments
         a = [0, 1, 2]
@@ -377,11 +381,12 @@ class ParquetTest(ArkoudaTest):
             ak.array([0, 0, len(a), len(a), len(a), len(a) + len(c)]), ak.array(a + c, dtype=ak.uint64)
         )
         with tempfile.TemporaryDirectory(dir=ParquetTest.par_test_base_tmp) as tmp_dirname:
-            s.to_parquet(f"{tmp_dirname}/int_test")
+            s.to_parquet(f"{tmp_dirname}/uint_test_empty")
 
-            rd_data = ak.read_parquet(f"{tmp_dirname}/int_test*")
+            rd_data = ak.read_parquet(f"{tmp_dirname}/uint_test_empty*")
             for i in range(6):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
+        print("Finished UInt Empty")
 
         # bool test
         a = [0, 1, 1]
@@ -389,11 +394,12 @@ class ParquetTest(ArkoudaTest):
         c = [1, 0]
         s = ak.SegArray(ak.array([0, len(a), len(a) + len(b)]), ak.array(a + b + c, dtype=ak.bool))
         with tempfile.TemporaryDirectory(dir=ParquetTest.par_test_base_tmp) as tmp_dirname:
-            s.to_parquet(f"{tmp_dirname}/int_test")
+            s.to_parquet(f"{tmp_dirname}/bool_test")
 
-            rd_data = ak.read_parquet(f"{tmp_dirname}/int_test*")
+            rd_data = ak.read_parquet(f"{tmp_dirname}/bool_test*")
             for i in range(3):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
+        print("Finished Bool")
 
         # bool with empty segments
         a = [0, 1, 1]
@@ -402,11 +408,12 @@ class ParquetTest(ArkoudaTest):
             ak.array([0, 0, len(a), len(a), len(a), len(a) + len(c)]), ak.array(a + c, dtype=ak.bool)
         )
         with tempfile.TemporaryDirectory(dir=ParquetTest.par_test_base_tmp) as tmp_dirname:
-            s.to_parquet(f"{tmp_dirname}/int_test")
+            s.to_parquet(f"{tmp_dirname}/bool_test_empty")
 
-            rd_data = ak.read_parquet(f"{tmp_dirname}/int_test*")
+            rd_data = ak.read_parquet(f"{tmp_dirname}/bool_test_empty*")
             for i in range(6):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
+        print("Finished Bool Empty")
 
         # float test
         a = [1.1, 1.1, 2.7]
@@ -414,22 +421,24 @@ class ParquetTest(ArkoudaTest):
         c = [15.2, 21.0]
         s = ak.SegArray(ak.array([0, len(a), len(a) + len(b)]), ak.array(a + b + c))
         with tempfile.TemporaryDirectory(dir=ParquetTest.par_test_base_tmp) as tmp_dirname:
-            s.to_parquet(f"{tmp_dirname}/int_test")
+            s.to_parquet(f"{tmp_dirname}/float_test")
 
-            rd_data = ak.read_parquet(f"{tmp_dirname}/int_test*")
+            rd_data = ak.read_parquet(f"{tmp_dirname}/float_test*")
             for i in range(3):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
+        print("Finished Float")
 
         # float with empty segments
         a = [1.1, 1.1, 2.7]
         c = [15.2, 21.0]
         s = ak.SegArray(ak.array([0, 0, len(a), len(a), len(a), len(a) + len(c)]), ak.array(a + c))
         with tempfile.TemporaryDirectory(dir=ParquetTest.par_test_base_tmp) as tmp_dirname:
-            s.to_parquet(f"{tmp_dirname}/int_test")
+            s.to_parquet(f"{tmp_dirname}/float_test_empty")
 
-            rd_data = ak.read_parquet(f"{tmp_dirname}/int_test*")
+            rd_data = ak.read_parquet(f"{tmp_dirname}/float_test_empty*")
             for i in range(6):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
+        print("Finished Float Empty")
 
     def test_multicol_write(self):
         df_dict = {

--- a/tests/parquet_test.py
+++ b/tests/parquet_test.py
@@ -372,8 +372,6 @@ class ParquetTest(ArkoudaTest):
             for i in range(3):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
 
-        print("Finished UInt")
-
         # uint with empty segments
         a = [0, 1, 2]
         c = [15, 21]
@@ -386,7 +384,6 @@ class ParquetTest(ArkoudaTest):
             rd_data = ak.read_parquet(f"{tmp_dirname}/uint_test_empty*")
             for i in range(6):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
-        print("Finished UInt Empty")
 
         # bool test
         a = [0, 1, 1]
@@ -399,7 +396,6 @@ class ParquetTest(ArkoudaTest):
             rd_data = ak.read_parquet(f"{tmp_dirname}/bool_test*")
             for i in range(3):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
-        print("Finished Bool")
 
         # bool with empty segments
         a = [0, 1, 1]
@@ -413,7 +409,6 @@ class ParquetTest(ArkoudaTest):
             rd_data = ak.read_parquet(f"{tmp_dirname}/bool_test_empty*")
             for i in range(6):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
-        print("Finished Bool Empty")
 
         # float test
         a = [1.1, 1.1, 2.7]
@@ -426,7 +421,6 @@ class ParquetTest(ArkoudaTest):
             rd_data = ak.read_parquet(f"{tmp_dirname}/float_test*")
             for i in range(3):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
-        print("Finished Float")
 
         # float with empty segments
         a = [1.1, 1.1, 2.7]
@@ -438,7 +432,6 @@ class ParquetTest(ArkoudaTest):
             rd_data = ak.read_parquet(f"{tmp_dirname}/float_test_empty*")
             for i in range(6):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
-        print("Finished Float Empty")
 
     def test_multicol_write(self):
         df_dict = {

--- a/tests/parquet_test.py
+++ b/tests/parquet_test.py
@@ -346,7 +346,6 @@ class ParquetTest(ArkoudaTest):
             rd_data = ak.read_parquet(f"{tmp_dirname}/int_test*")
             for i in range(3):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
-        print("Finished Int")
 
         # integer with empty segments
         a = [0, 1, 2]
@@ -358,7 +357,6 @@ class ParquetTest(ArkoudaTest):
             rd_data = ak.read_parquet(f"{tmp_dirname}/int_test_empty*")
             for i in range(6):
                 self.assertListEqual(s[i].to_list(), rd_data[i].to_list())
-        print("Finished Int Empty")
 
         # uint test
         a = [0, 1, 2]

